### PR TITLE
saltpack: decryption U/I for sender identification

### PR DESCRIPTION
**Pull request by maxtaco**
- decryption UI plumbed through
- clean up use of identify UI
  - change the Start() function to say why the identification is happening. We're going to
    need this anyways, for KBFS integrations.
- fix the blocking semantics needed for id2 use in --interactive mode
- more tests for decryption conditions

### CODE REVIEW

**Comment by maxtaco** [comment]
@patrickxb can you take a look?

File: go/engine/saltpack_decrypt.go [hunk 1]
```diff
@@ -40,21 +43,95 @@ func (e *SaltPackDecrypt) Prereqs() Prereqs {
...
+	switch eng.getTrackType() {
+	case identify2NoTrack:
+		arg.SenderType = keybase1.SaltPackSenderType_NOT_TRACKED
+	case identify2TrackOK:
+		arg.SenderType = keybase1.SaltPackSenderType_TRACKING_OK
+	default:
+		arg.SenderType = keybase1.SaltPackSenderType_TRACKING_BROKE
+	}
+	return nil
+}
```
Comment (patrickxb) [inline go/engine/saltpack_decrypt.go]: Can this function be moved somewhere usable by other engines?  Verify will need to do the same thing.

File: go/libkb/saltpack_dec.go [hunk 2]
```diff
@@ -11,13 +11,19 @@ import (
 
 func SaltPackDecrypt(
 	source io.Reader, sink io.WriteCloser,
-	deviceEncryptionKey NaclDHKeyPair) error {
-	_, plainsource, frame, err := saltpack.NewDearmor62DecryptStream(
+	deviceEncryptionKey NaclDHKeyPair, checkSender func(*saltpack.MessageKeyInfo) error) error {
```
Comment (patrickxb) [inline go/libkb/saltpack_dec.go]: This 80-char wrapping seems particularly unreadable...perhaps it looks ok on punch cards, but...

File: go/libkb/util.go [hunk 3]
```diff
@@ -414,6 +414,11 @@ func Trace(log logger.Logger, msg string, f func() error) func() {
 	return func() { log.Debug("- %s -> %s", msg, ErrToOk(f())) }
 }
 
+func (g *GlobalContext) Trace(msg string, f func() error) func() {
```
Comment (patrickxb) [inline go/libkb/util.go]: change to call Trace function above:

    return Trace(g.Log, msg, f)

**Comment by patrickxb** [comment]
If identifySender can be moved somewhere another engine can use it (i.e. verify), then LGTM.  Other comments are minor.

File: go/engine/saltpack_decrypt.go (hunk 1 above)
Comment (maxtaco) [inline go/engine/saltpack_decrypt.go]: how about the related functions?  Maybe you want everything from `computeSenderArg` on down.

Comment (patrickxb) [inline go/engine/saltpack_decrypt.go]: verify doesn't use MessageKeyInfo

Comment (maxtaco) [inline go/engine/saltpack_decrypt.go]: i'll factor it out.

Comment (patrickxb) [inline go/engine/saltpack_decrypt.go]: So the lookupSender wouldn't work...different key type.

Comment (maxtaco) [inline go/engine/saltpack_decrypt.go]: still lookup by public key, right?

Comment (maxtaco) [inline go/engine/saltpack_decrypt.go]: Ah, no.  We can change it to lookupByKeybaseKID.

Comment (patrickxb) [inline go/engine/saltpack_decrypt.go]: ok, sounds good.

Comment (maxtaco) [inline go/engine/saltpack_decrypt.go]: We'll need a way to coerce sign/verify saltpack keys into `keybase1.KID`s, but that should be OK, right?

Comment (patrickxb) [inline go/engine/saltpack_decrypt.go]: They all implement KIDExtractor, so they all have ToKID().

**Comment by maxtaco** [comment]
Ready for another lookie!

**Comment by patrickxb** [comment]
LGTM

**Pull request merged**